polygon/heimdall/heimdall.go

```diff
@@ -82,6 +82,13 @@ func (h *heimdall) FetchCheckpointsFromBlock(ctx context.Context, startBlock uin
 		if err != nil {
 			return nil, err
 		}
+		if len(checkpoints) == 0 {
+			return nil, errors.New("unexpected empty checkpoints")
+		}
+		if checkpoints[len(checkpoints)-1].CmpRange(startBlock) > 0 {
+			// the start block is past the last checkpoint
+			return nil, nil
+		}
```

Comment:
I assume it fixes this, correct?

> when the tip is after the latest checkpoint.EndBlock we returned error which caused the sync to be stuck - we should return empty list

why is it happening? why the sync is stuck?
do you think this logic fits best fetcher or sync itself?

if you think it fits the fetcher, we need to update the entityFetcher impl to have this too

P.S. the heimdall.go is obsolete and my plan was to switch to heimdall/service.go and delete heimdall.go as soon as there's an integration test for it. I got stuck with an integration test (it is more work than I thought), so I'm also considering doing it already since you are testing it manually already. It is a pity that the lack of automated tests blocks manual testing.

Reply:
@battlmonstr this is because the latest checkpoint EndBlock is always behind the current tip header. You can check this by going to:
- https://heimdall-api-amoy.polygon.technology/checkpoints/latest and look at the end block number (at time of writing 7853938)
- https://amoy.polygonscan.com and look at the latest block number (at time of writing 7859244)

So if syncing starts at 7859244-20 as an example it needs to exit `DownloadBlocksUsingCheckpoints(start)` without producing a new tip so it can then proceed to `DownloadBlocksUsingMilestones(start)`.

To answer your other point, `heimdall.go` is not obsolete yet because it supports snapshots. We can't switch to using `heimdall/service.go` until it supports snapshots.

Reply:
As to updating the `entityFetcherImpl` do you mind adding that to your list? I'm focusing on getting the stage to work at the moment.